File: OpenWA/sdk/php/src/Resources/ChannelsResource.php
<?php

declare(strict_types=1);

namespace OpenWA\Resources;

use OpenWA\Http\HttpExecutor;

/**
 * Channels resource — WhatsApp Channels / Newsletters.
 *
 * Backed by src/modules/channel/channel.controller.ts (@Controller('sessions/:sessionId/channels')).
 */
class ChannelsResource
{
    private HttpExecutor $http;

    public function __construct(HttpExecutor $http)
    {
        $this->http = $http;
    }

    /** @return array<int,array<string,mixed>> */
    public function list(string $sessionId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/channels") ?? [];
    }

    /** @return array<string,mixed> */
    public function get(string $sessionId, string $channelId): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/channels/{$this->http->encodeSegment($channelId)}");
    }

    /**
     * Get recent messages from a channel.
     *
     * @param array<string,mixed> $query  e.g. ['limit' => 50].
     * @return array<int,array<string,mixed>>
     */
    public function messages(string $sessionId, string $channelId, array $query = []): array
    {
        return $this->http->request('GET', "/api/sessions/{$this->http->encodeSegment($sessionId)}/channels/{$this->http->encodeSegment($channelId)}/messages", $query) ?? [];
    }

    /**
     * Create a channel. The account owns it, which is what makes delete() possible later.
     *
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    public function create(string $sessionId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/channels", [], $body);
    }

    /**
     * Delete a channel this account owns. Irreversible; every subscriber loses it.
     *
     * Note the path: unsubscribe is the DELETE route, and the two are deliberately not reachable by
     * the same request — leaving a channel and destroying it are very different acts.
     *
     * @return array<string,mixed>
     */
    public function delete(string $sessionId, string $channelId): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/channels/{$this->http->encodeSegment($channelId)}/delete");
    }

    /**
     * Mute or unmute a channel's notifications. The subscription is untouched either way.
     *
     * @param array<string,mixed> $body
     * @return array<string,mixed>
     */
    public function mute(string $sessionId, string $channelId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/channels/{$this->http->encodeSegment($channelId)}/mute", [], $body);
    }
    /**
     * Subscribe to a channel using its invite code. Requires an OPERATOR-level key.
     *
     * @param array<string,mixed> $body  Must contain 'inviteCode'.
     * @return array<string,mixed>
     */
    public function subscribe(string $sessionId, array $body): array
    {
        return $this->http->request('POST', "/api/sessions/{$this->http->encodeSegment($sessionId)}/channels/subscribe", [], $body);
    }

    /** Unsubscribe from a channel. Requires an OPERATOR-level key. */
    public function unsubscribe(string $sessionId, string $channelId): array
    {
        return $this->http->request('DELETE', "/api/sessions/{$this->http->encodeSegment($sessionId)}/channels/{$this->http->encodeSegment($channelId)}");
    }
}

File: OpenWA/sdk/java/src/main/java/com/rmyndharis/openwa/resources/ChannelsResource.java
package com.rmyndharis.openwa.resources;

import static com.rmyndharis.openwa.http.Http.encodeSegment;

import com.rmyndharis.openwa.OpenWAClient;
import com.rmyndharis.openwa.http.HttpMethod;
import com.rmyndharis.openwa.model.ChannelMessageQuery;
import com.rmyndharis.openwa.model.ChannelMessageRecord;
import com.rmyndharis.openwa.model.ChannelRecord;
import com.rmyndharis.openwa.model.CreateChannelRequest;
import com.rmyndharis.openwa.model.MuteChannelRequest;
import com.rmyndharis.openwa.model.SubscribeChannelRequest;
import com.rmyndharis.openwa.model.SuccessResult;
import java.util.List;

/**
 * Channels resource — WhatsApp Channels / Newsletters.
 *
 * <p>Backed by {@code sessions/:sessionId/channels}.
 */
public final class ChannelsResource {
    private final OpenWAClient client;

    public ChannelsResource(OpenWAClient client) {
        this.client = client;
    }

    /** List all channels/newsletters the session is subscribed to. */
    public List<ChannelRecord> list(String sessionId) {
        return client.requestList(
            HttpMethod.GET, "/api/sessions/" + encodeSegment(sessionId) + "/channels", null, null, ChannelRecord.class);
    }

    /** Get a single channel by id. */
    public ChannelRecord get(String sessionId, String channelId) {
        return client.request(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/channels/" + encodeSegment(channelId),
            null,
            null,
            ChannelRecord.class);
    }

    /** Get recent messages from a channel. */
    public List<ChannelMessageRecord> messages(String sessionId, String channelId, ChannelMessageQuery query) {
        return client.requestList(
            HttpMethod.GET,
            "/api/sessions/" + encodeSegment(sessionId) + "/channels/" + encodeSegment(channelId) + "/messages",
            query,
            null,
            ChannelMessageRecord.class);
    }

    /** Create a channel. The account owns it, which is what makes {@link #delete} possible later. */
    public ChannelRecord create(String sessionId, CreateChannelRequest body) {
        return client.request(
            HttpMethod.POST, "/api/sessions/" + encodeSegment(sessionId) + "/channels", null, body, ChannelRecord.class);
    }

    /**
     * Delete a channel this account owns. Irreversible, and every subscriber loses it.
     *
     * <p>Note the path: {@code unsubscribe} is the {@code DELETE} route, and the two are deliberately
     * not reachable by the same request — leaving a channel and destroying it are very different acts.
     */
    public SuccessResult delete(String sessionId, String channelId) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/channels/" + encodeSegment(channelId) + "/delete",
            null,
            null,
            SuccessResult.class);
    }

    /** Mute or unmute a channel's notifications. The subscription is untouched either way. */
    public SuccessResult mute(String sessionId, String channelId, MuteChannelRequest body) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/channels/" + encodeSegment(channelId) + "/mute",
            null,
            body,
            SuccessResult.class);
    }

    /** Subscribe to a channel using its invite code. Requires an OPERATOR-level key. */
    public ChannelRecord subscribe(String sessionId, SubscribeChannelRequest body) {
        return client.request(
            HttpMethod.POST,
            "/api/sessions/" + encodeSegment(sessionId) + "/channels/subscribe",
            null,
            body,
            ChannelRecord.class);
    }

    /** Unsubscribe from a channel. Requires an OPERATOR-level key. */
    public SuccessResult unsubscribe(String sessionId, String channelId) {
        return client.request(
            HttpMethod.DELETE,
            "/api/sessions/" + encodeSegment(sessionId) + "/channels/" + encodeSegment(channelId),
            null,
            null,
            SuccessResult.class);
    }
}
